tests/unit/nupic/support/decorators_test.py

```diff
@@ -22,110 +22,345 @@
+      timeoutSec=1, initialRetryDelaySec=1,
+      maxRetryDelaySec=10, retryExceptions=(TestParentException,),
+      retryFilter=lambda _1, _2, _3: True)
+
+    @retryDecoratorTrueFilter
+    def testFunctionTrue():
+      raise TestChildException("Test exception")
+
+    with self.assertRaises(TestChildException):
+      testFunctionTrue()
+
+    self.assertEqual(mockSleep.call_count, 1)
+
+    # Test with retryFilter returning False
+
+    mockSleep.reset_mock()
+
+    retryDecoratorFalseFilter = decorators.retry(
+      timeoutSec=1, initialRetryDelaySec=1,
+      maxRetryDelaySec=10, retryExceptions=(TestParentException,),
+      retryFilter=lambda _1, _2, _3: False)
+
+    @retryDecoratorFalseFilter
+    def testFunctionFalse():
+      raise TestChildException("Test exception")
+
+    with self.assertRaises(TestChildException):
+      testFunctionFalse()
+
+    self.assertEqual(mockSleep.call_count, 0)
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testReturnsExpectedWithExpectedArgs(self, mockTime, mockSleep):
+    """Test that docorated function receives only expected args and
+    that it returns the expected value on success."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    retryDecorator = decorators.retry(
+      timeoutSec=30, initialRetryDelaySec=2,
+      maxRetryDelaySec=10)
+
+    testFunction = Mock(return_value=321,
+                        __name__="testFunction", autospec=True)
+
+    returnValue = retryDecorator(testFunction)(1, 2, a=3, b=4)
+
+    self.assertEqual(returnValue, 321)
+    testFunction.assert_called_once_with(1, 2, a=3, b=4)
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testNoRetryIfCallSucceeds(self, mockTime, mockSleep):
+    """If the initial call succeeds, test that no retries are performed."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    retryDecorator = decorators.retry(
+      timeoutSec=30, initialRetryDelaySec=2,
+      maxRetryDelaySec=10)
+
+    testFunction = Mock(__name__="testFunction", autospec=True)
+
+    retryDecorator(testFunction)()
+
+    testFunction.assert_called_once_with()
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testFailsFirstSucceedsLater(self, mockTime, mockSleep):
+    """If initial attempts fail but subsequent attempt succeeds, ensure that
+    expected number of retries is performed and expected value is returned."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    retryDecorator = decorators.retry(
+      timeoutSec=30, initialRetryDelaySec=2,
+      maxRetryDelaySec=10)
+
+    testFunction = Mock(
+      side_effect=[
+        TestParentException("Test exception 1"),
+        TestParentException("Test exception 2"),
+        321
+      ],
+      __name__="testFunction", autospec=True)
+
+    returnValue = retryDecorator(testFunction)()
+
+    self.assertEqual(returnValue, 321)
+    self.assertEqual(testFunction.call_count, 3)
```

Comment:
Just wanted to remind about adding test(s) for the logging arg to the retry decorator. Both passed and not passed.